pkg/infrastructure/azure/dns.go

```diff
@@ -0,0 +1,242 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/dns/armdns"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/privatedns/armprivatedns"
+	"github.com/sirupsen/logrus"
+	"k8s.io/utils/ptr"
+	capz "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	"github.com/openshift/installer/pkg/infrastructure/clusterapi"
+	"github.com/openshift/installer/pkg/types"
+)
+
+type recordListType string
+
+const (
+	// ExternalPublishingStrategy exposes endpoints for the cluster to the Internet.
+	cname recordListType = "Cname"
+	// InternalPublishingStrategy exposes the endpoints for the cluster to the private network only.
+	arecord recordListType = "ARecord"
+	// MixedPublishingStrategy allows for the api server and the ingress to be configured individually for exposure to
+	// private network or Internet.
+	aaaarecord recordListType = "AaaaRecord"
+)
+
+type recordList struct {
+	Name       string
+	RecordType armdns.RecordType
+	RecordSet  armdns.RecordSet
+}
+
+type recordPrivateList struct {
+	Name       string
+	RecordType armprivatedns.RecordType
+	RecordSet  armprivatedns.RecordSet
+}
+
+// Create DNS entries for azure.
+func createDNSEntries(ctx context.Context, in clusterapi.InfraReadyInput) error {
+	infraID := in.InfraID
+	private := in.InstallConfig.Config.Publish == types.InternalPublishingStrategy
+	baseDomainResourceGroup := in.InstallConfig.Config.Azure.BaseDomainResourceGroupName
+	resourceGroup := fmt.Sprintf("%s-rg", in.InfraID)
+	zone := in.InstallConfig.Config.BaseDomain
+	privatezone := in.InstallConfig.Config.ClusterDomain()
+	apiExternalName := fmt.Sprintf("api.%s", infraID)
+	apiExternalNameV6 := fmt.Sprintf("v6-api.%s", infraID)
+	// TODO: set value from either manifests or getting from client.
+	azureTags := make(map[string]*string)
+	for k, v := range in.InstallConfig.Config.Azure.UserTags {
+		azureTags[k] = ptr.To(v)
+	}
+	azureCluster := &capz.AzureCluster{}
+	key := client.ObjectKey{
+		Name:      in.InfraID,
+		Namespace: capiutils.Namespace,
+	}
+	if err := in.Client.Get(ctx, key, azureCluster); err != nil {
+		return fmt.Errorf("failed to get Azure cluster: %w", err)
+	}
+
+	ipIlb := azureCluster.Spec.NetworkSpec.APIServerLB.FrontendIPs[0].PrivateIPAddress
+
+	// TODO: Populate with IP after external LB is created.
+	var ipElb string
+
+	useIPv6 := false
+	for _, network := range in.InstallConfig.Config.Networking.ServiceNetwork {
+		if network.IP.To4() == nil {
+			useIPv6 = true
+		}
+	}
+
+	privateRecords := []recordPrivateList{}
+	ttl := int64(300)
+	recordType := arecord
+	if useIPv6 {
+		recordType = aaaarecord
+	}
+	publicRecords := createRecordSet(fmt.Sprintf("api.%s", in.InstallConfig.Config.ObjectMeta.Name), azureTags, ttl, recordType, ipElb, "")
```

Comment:
why have public records here and then overwrite it in the `if !private` condition below? do we use these public records if `private`?